File: packages/typescript/src/runner.ts
import { EvalCase, EvalReport, EvalResult } from "./models";
import { Scorer } from "./scorers/base";

export type ModelFn = (input: string) => Promise<string> | string;

export interface RunOptions {
  passThreshold?: number;
  concurrency?:   number;
}

export async function runEval(
  cases:    EvalCase[],
  scorer:   Scorer,
  opts:     RunOptions & ({ modelFn: ModelFn; outputs?: never } | { outputs: string[]; modelFn?: never }),
): Promise<EvalReport> {
  const threshold = opts.passThreshold ?? 0.5;
  let outputs: string[];

  if ("outputs" in opts && opts.outputs !== undefined) {
    if (opts.outputs.length !== cases.length) {
      throw new Error(`outputs length ${opts.outputs.length} !== cases length ${cases.length}`);
    }
    outputs = opts.outputs;
  } else if ("modelFn" in opts && opts.modelFn !== undefined) {
    outputs = await runParallel(cases.map((c) => c.input), opts.modelFn, opts.concurrency ?? 4);
  } else {
    throw new Error("Provide either modelFn or outputs");
  }

  const results: EvalResult[] = cases.map((c, i) => {
    const r = scorer.score(c, outputs[i]);
    r.passed = r.score >= threshold;
    return r;
  });

  return { results, scorerName: scorer.name, passThreshold: threshold };
}

export async function compare(
  cases:     EvalCase[],
  scorer:    Scorer,
  baseline:  ModelFn,
  candidate: ModelFn,
  opts:      RunOptions = {},
): Promise<[EvalReport, EvalReport]> {
  const [baseOut, candOut] = await Promise.all([
    runParallel(cases.map((c) => c.input), baseline,  opts.concurrency ?? 4),
    runParallel(cases.map((c) => c.input), candidate, opts.concurrency ?? 4),
  ]);
  const [baseReport, candReport] = await Promise.all([
    runEval(cases, scorer, { outputs: baseOut, passThreshold: opts.passThreshold }),
    runEval(cases, scorer, { outputs: candOut, passThreshold: opts.passThreshold }),
  ]);
  return [baseReport, candReport];
}

async function runParallel(inputs: string[], fn: ModelFn, concurrency: number): Promise<string[]> {
  const results: string[] = new Array(inputs.length);
  for (let i = 0; i < inputs.length; i += concurrency) {
    const batch   = inputs.slice(i, i + concurrency);
    const outputs = await Promise.all(batch.map((inp) => Promise.resolve(fn(inp))));
    outputs.forEach((out, j) => { results[i + j] = out; });
  }
  return results;
}

File: packages/python/evalforge/runner.py
from __future__ import annotations
from typing import Callable
from concurrent.futures import ThreadPoolExecutor, as_completed
from .models import EvalCase, EvalDataset, EvalReport
from .scorers.base import Scorer


ModelFn = Callable[[str], str]


def run_eval(
    dataset:        EvalDataset,
    scorer:         Scorer,
    model_fn:       ModelFn | None    = None,
    outputs:        list[str] | None  = None,
    pass_threshold: float             = 0.5,
    max_workers:    int               = 4,
) -> EvalReport:
    """Run an evaluation.

    Supply either `model_fn` (called per case to generate outputs) or
    pre-generated `outputs` (list aligned with dataset.cases).
    """
    if model_fn is None and outputs is None:
        raise ValueError("Provide either model_fn or outputs")
    if outputs is not None and len(outputs) != len(dataset.cases):
        raise ValueError(f"outputs length {len(outputs)} != dataset length {len(dataset)}")

    if outputs is None:
        outputs = _generate(dataset.cases, model_fn, max_workers)  # type: ignore[arg-type]

    results = [scorer.score(case, out) for case, out in zip(dataset.cases, outputs)]

    # Respect pass_threshold override per report (individual scorers set their own internally)
    for r in results:
        r.passed = r.score >= pass_threshold

    return EvalReport(
        results=results,
        scorer_name=scorer.name,
        pass_threshold=pass_threshold,
    )


def compare(
    dataset:   EvalDataset,
    scorer:    Scorer,
    baseline:  ModelFn,
    candidate: ModelFn,
    max_workers: int = 4,
) -> tuple[EvalReport, EvalReport]:
    """Run eval on two model functions and return (baseline_report, candidate_report)."""
    base_outputs = _generate(dataset.cases, baseline,  max_workers)
    cand_outputs = _generate(dataset.cases, candidate, max_workers)
    base_report  = run_eval(dataset, scorer, outputs=base_outputs)
    cand_report  = run_eval(dataset, scorer, outputs=cand_outputs)
    return base_report, cand_report


def _generate(cases: list[EvalCase], fn: ModelFn, max_workers: int) -> list[str]:
    outputs: list[str] = [""] * len(cases)
    with ThreadPoolExecutor(max_workers=max_workers) as pool:
        futures = {pool.submit(fn, c.input): i for i, c in enumerate(cases)}
        for fut in as_completed(futures):
            outputs[futures[fut]] = fut.result()
    return outputs
